src/contract_client.rs

```diff
@@ -804,24 +800,208 @@ pub fn decode_concordium_std_error(reject_reason: i32) -> Option<ConcordiumStdRe
+/// but not the `return_value`. The reason can be explained by the historical
+/// development of smart contracts on Concordium (types were not expanded to
+/// avoid breaking changes) as well as to keep the size of the node as small as
+/// possible since the `return_value` does not need to be saved by the node for
+/// achieving consensus.
+///
+/// As a result, this decoding function needs to have access to both the
+/// `return_value` and the `error_code` to decode the `reject_reason` of a
+/// reverted transaction into a human-readable error.
+///
+/// How is the `return_value` and `error_code` assigned in rejected
+/// transactions:
+/// - If the transaction reverts due to an error in the `concordium-std` crate,
+/// the `return_value` is None and the `error_code` is assigned as defined in
+/// the `concordium-std` crate.
+/// - If the transaction reverts due to an error in the smart contract logic:
+/// A smart contract V1 needs to implement a conversion to `Reject` for its
+/// smart contract errors.
+/// `<https://docs.rs/concordium-std/latest/concordium_std/struct.Reject.html>`
+///
+/// 1. Example: Deriving `Reject` in the smart contract.
+/// The simplest way to implement `Reject` in the smart contract is by deriving
+/// it.
+///
+/// ```ignore
+/// /// The custom errors the contract can produce.
+/// #[derive(Serialize, Debug, Reject, SchemaType)]
+/// pub enum CustomContractError {
+///     /// CustomError1 defined in the smart contract logic.
+///     CustomError1, // return_value: 00; error_code: -1
+///     /// CustomError2 defined in the smart contract logic.
+///     CustomError2, // return_value: 01; error_code: -2
+/// }
+/// ```
+///
+/// The `Reject` macro assigns the `error_codes` starting from `-1` to the enum
+/// variants and assigns the `return_values` by serializing the enum variants.
+/// The `return_values` are equivalent to the enum tags in the above example.
+///
+/// The JSON value returned by this function for the above `CustomError1` is:
+/// ```json
+/// {"CustomError1":[]}
+/// ```
+///
+/// 2. Example: Deriving `Reject` in the smart contract with nested errors.
+/// Nested errors are often used to inherit the errors from a smart
+/// contract library such as the cis2-library.
+/// `<https://github.com/Concordium/concordium-rust-smart-contracts/blob/dde42fa62254a55b46a4c9c52c32bbe661127001/concordium-cis2/src/lib.rs#L1093>`
+///
+/// Library:
+/// ```ignore
+/// /// The custom errors the library can produce.
+/// #[derive(Serialize, Debug, Reject, SchemaType)]
+/// pub enum Cis2Error<R> {
+///     /// CustomErrorLibrary1 defined in the smart contract logic.
+///     CustomErrorLibrary1, // return_value: 00; error_code: -1
+///     /// Nested error variant.
+///     Custom(R),
+///     // Custom::CustomError1 -> return_value: 0100; error_code: -1
+///     // Custom::CustomError2 -> return_value: 0101; error_code: -2
+///     // Custom::CustomError3 -> return_value: 0102; error_code: -3
```

Comment:
```suggestion
///     // CustomContractError::CustomError1 -> return_value: 0100; error_code: -1
///     // CustomContractError::CustomError2 -> return_value: 0101; error_code: -2
///     // CustomContractError::CustomError3 -> return_value: 0102; error_code: -3
```

Right?

Reply:
I changed it to 
```
///     // CustomError1 -> return_value: 0100; error_code: -1
///     // CustomError2 -> return_value: 0101; error_code: -2
///     // CustomError3 -> return_value: 0102; error_code: -3
```

I used `Custom::CustomError1` because that is how the error is displayed with the schema, but I see that it is confusing because in code you would address the error variant as `CustomContractError::CustomError1`.